Leverage Intl.DateTimeFormat for time zone parsing

josefarias opened the pull request:
I wanna clean this up a bit before putting up for review

josefarias commented on lib/assets/javascripts/src/local-time/helpers/strftime.coffee:
```diff
@@ -2,6 +2,13 @@ import LocalTime from "../local_time"
 
+supportsIntlDateFormat = typeof Intl?.DateTimeFormat is "function"
+
+knownEdgeCaseTimeZones =
```
https://github.com/basecamp/local_time/pull/141#issuecomment-1646775570

josefarias commented on lib/assets/javascripts/src/local-time/helpers/strftime.coffee:
```diff
@@ -40,6 +50,23 @@ pad = (num, flag) ->
+  Object.keys(knownEdgeCaseTimeZones).find (name) ->
+    time.toString().indexOf(name) isnt -1
+
+parseTimeZoneWithIntl = (time) ->
+  return null unless supportsIntlDateFormat
+  result = new Date(time).toLocaleString(undefined, { timeZoneName: "short" }).split(" ").pop()
+  result if result.indexOf("GMT") is -1
```
`Date.toLocaleString()` uses `Intl` behind the scenes: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString

It will return `GMT+/-X` when it fails to return a time zone abbreviation.

josefarias commented:
@ramhoj from https://3.basecamp.com/2914079/buckets/27/card_tables/cards/5644726248#__recording_5645012094 it seems that Swedish time zones are also incorrect on `main`. If so, could I get the expected values please, so that I can make sure this PR also fixes those?

josefarias commented on lib/assets/javascripts/src/local-time/helpers/strftime.coffee:
```diff
@@ -40,6 +50,23 @@ pad = (num, flag) ->
+
+edgeCaseTimeZone = (time) ->
+  Object.keys(knownEdgeCaseTimeZones).find (name) ->
+    time.toString().indexOf(name) isnt -1
```
We have to match on the long time zone name because there are some geographies which share the same abbreviation for different time zones. From ChatGPT (which wouldn't give me an exhaustive list):

* PST:
  * Pacific Standard Time (UTC-8): observed in parts of the United States and Canada.
  * Philippine Time (UTC+8): observed in the Philippines.
* EST:
  * Eastern Standard Time (UTC-5): observed in parts of the United States and Canada.
  * Eastern Standard Time (UTC+10): observed in parts of Australia during their standard time.
  * _Note: glad we don't have to match on this one since they even have the same long name_
* MST:
  * Mountain Standard Time (UTC-7): observed in parts of the United States and Canada.
  * Malaysia Standard Time (UTC+8): observed in Malaysia.
* IST:
  * Indian Standard Time (UTC+5:30): observed in India.
  * Irish Standard Time (UTC+1): observed in Ireland during standard time (UTC+0 during daylight saving time).
* CAT:
  * Central Africa Time (UTC+2): observed in several African countries.
  * Central Alaska Time (UTC-9): observed in some areas of Alaska, USA.
* AST:
  * Atlantic Standard Time (UTC-4): observed in parts of Canada and the Caribbean.
  * Arabia Standard Time (UTC+3): observed in the Arabian Peninsula.
* GMT:
  * Greenwich Mean Time (UTC+0): observed in the United Kingdom and other regions.
  * Gambia Mean Time (UTC+0): observed in The Gambia.
* HST:
  * Hawaii Standard Time (UTC-10): observed in Hawaii, USA.
  * Honduras Standard Time (UTC-6): observed in Honduras.
* CST:
  * Central Standard Time (UTC-6): observed in parts of North America, including parts of the United States and Canada.
  * China Standard Time (UTC+8): observed in China and other regions.

afcapel replied:
Can we use `String.prototype.includes` here?

josefarias commented on lib/assets/javascripts/src/local-time/helpers/strftime.coffee:
```diff
@@ -2,6 +2,16 @@ import LocalTime from "../local_time"
+  "Western Indonesia Time": "WIB"
+  "Singapore Standard Time": "SGT"
+  "Москва, стандартное время": "MSK" # Europe/Moscow
+  "中国标准时间": "CST" # Asia/Shanghai
```
This is the name of the timezone as returned in my javascript console. 

Notice how some time zone names in this list are in english, while others aren't. I just realized this might be due to only some of my chrome sensor locations including language locale information. The locations that don't include a locale are probably defaulting to english.

I'll test this out tomorrow. But this might mean that the browser's language might interfere with matching on these names.

I hope I'm wrong. And if I'm not then I hope there's a reliable way to get the time zone name in the same language every time.

josefarias replied:
Did a super quick test that I'll need to validate later but this seems fixable!
![CleanShot 2023-08-21 at 11 35 46](https://github.com/basecamp/local_time/assets/31393016/66f36a80-54fc-4109-8b7a-0f4362aed2bc)

ramhoj replied:
Ohh, that's super nice, good catch!

ramhoj replied:
Ohh, that's really good for detecting the special cases, great catch!

josefarias replied:
This seems to have worked!

ramhoj commented on CONTRIBUTING.md:
```diff
@@ -10,9 +10,33 @@ Then, open your console on the project directory and run `bundle install` and `y
+Pacific/Honolulu  | HST                   | —                | 21.5376  | -158.0023
+Asia/Singapore    | SGT                   | —                | 1.3521   | 103.8198
+Europe/Moscow     | MSK                   | —                | 55.7558  | 37.6173
+Asia/Shanghai     | CST                   | —                | 31.2304  | 121.4737
```
Great guide! Ideally the browser's Location settings would have an API we could use in our unit tests but this tooling is great even if we had that too.

ramhoj reviewed (commented):
Thank you @josefarias, localing timezone abbreviations proves super tricky if even possible!

Getting reasonable time zone abbreviations in all cases seems super tricky with how local_time is currently setup and I'm not sure how far we should attempt this here.

I suggested one approach in one of the review comments. Please have a look and see if it makes sense.

josefarias commented on lib/assets/javascripts/src/local-time/helpers/strftime.coffee:
```diff
@@ -40,6 +50,23 @@ pad = (num, flag) ->
+parseTimeZoneWithIntl = (time) ->
+  return null unless supportsIntlDateFormat
+  result = new Date(time).toLocaleString(undefined, { timeZoneName: "short" }).split(" ").pop()
+  result if result.indexOf("GMT") is -1
```
Yeah I'm not familiar with coffeescript idioms to be honest. Happy to change this in a future commit 👍

afcapel replied:
I think we can use `String.prototype.includes` here too.

josefarias commented on lib/assets/javascripts/src/local-time/helpers/strftime.coffee:
```diff
@@ -40,6 +50,23 @@ pad = (num, flag) ->
+
+parseTimeZoneWithIntl = (time) ->
+  return null unless supportsIntlDateFormat
+  result = new Date(time).toLocaleString(undefined, { timeZoneName: "short" }).split(" ").pop()
```
> To make matters worse the result differ if the user uses English in Jerusalem vs using Modern Hebrew. 

Yeah we'll have to find a fix for this, also confirmed in https://github.com/basecamp/local_time/pull/142#discussion_r1298101031. The [other solution](https://github.com/basecamp/local_time/pull/141) we had attempted I think is also liable to the same problem – but I'll have to check. Lastly, there are other local-time PRs that attempt to improve the heuristic method to include more time zones, but I agree we shouldn't keep leaning into that heuristic. I'll call this a "known unknown" and revisit once I come back to this PR next cooldown.

> I wonder if a better approach would be to skip the `parseTimeZoneWithHeuristic` all together and instead fallback on showing the GMT offset

This came up in https://github.com/basecamp/local_time/pull/141#issuecomment-1614649010, too. Tough to know what to do here. I had a quick talk about this with @rosa today and we agreed non-technical folk might not be familiar with GMT notation.

We've used the heuristic in production for many years and have only gotten reports for ~5 incorrect time zones that I can find. So I'd be comfortable with falling back to it after accounting for those edge-case time zones in `knownEdgeCaseTimeZones`. What do you think of this parsing sequence:

1. Check `knownEdgeCaseTimeZones` for the presence of the timezone (assuming we find a way to reliably get the same time zone name in english every time, per the beginning of this comment)
2. If not present, use `Intl` (i.e. via `Date.toLocaleString()`)
3. If we get back GMT notation, use the heuristic
4. If we get back an empty string (the heuristic's hard-coded fallback), print the GMT notation as a last resort
5. Finally, trust that we'll get reports for any incorrect results returned by the heuristic, and at that point we'll just add them to `knownEdgeCaseTimeZones`

I think that approach is, at the very least, better than what we have today. If this best-effort approach keeps falling short, we can always evaluate a different solution.

ramhoj replied:
I agree that we can react to bad abbreviations as they are reported. In most context the user will probably have a good idea of what the time zone is even if they wouldn't understand the zone abbreviation. The plan sounds solid and I agree that we shouldn't go for perfect here.

In the example section of the readme it looks like we're showing the time zone in the title attribute even if it's not included in the format string. What do you think of typing out the long time zone name in the title attribute (if it's an easy change)? Seems like a win even when we do have the right abbreviation but could dispel confusion for edge-cases we haven't caught yet?

josefarias replied:
> What do you think of typing out the long time zone name in the title attribute (if it's an easy change)

Oh that's interesting. Right, the title attribute doesn't obey the provided format in an effort to be the clearest possible (if I recall correctly). I'll look into the history of that when I come back to this PR. If that's the reasoning, then it'd make sense to use the long name there.

josefarias replied:
I've addressed this now, except for the part about changing the title which I'm gonna leave for another PR

afcapel reviewed (commented):
Looks great, thanks for tackling this @josefarias 

Do you think we can add some tests to try the new code paths?

josefarias commented:
Thanks for the reviews @ramhoj @afcapel. That was some great feedback. I've addressed it now.

> Do you think we can add some tests to try the new code paths?

Just added them 👍 I was hesitant to do so because, as you can see, it requires stubbing out some methods that don't belong to us. So we're operating in a highly fabricated environment.

I think the automated tests are okay to leave in for now to protect us from potential regressions. But we should be careful to always perform the manual tests described in `CONTRIBUTING.md`, too. Only the latter test "the real thing".

I've come to learn that stubbing out the browser's time zone requires quite a few gymnastics.

I'm looking to potentially merge this in tomorrow for testing in our own apps, looking to move towards a new release soon.

josefarias commented on test/javascripts/src/strftime_test.coffee:
```diff
@@ -28,6 +28,16 @@ momentMap =
   "%y": "YY"
   "%Y": "YYYY"
 
+stubToLocaleString = (stubImplementation) ->
+  original = Date.prototype.toLocaleString
+  Date.prototype.toLocaleString = stubImplementation
+  restore: -> Date.prototype.toLocaleString = original
+
+stubDateToString = (stubImplementation) ->
+  original = Date.prototype.toString
+  Date.prototype.toString = stubImplementation
+  restore: -> Date.prototype.toString = original
```
Slightly improved in https://github.com/basecamp/local_time/pull/154
